### PR TITLE
feat(ingest): Add accountType and employeeId for corpUser entity and ldap ingestion

### PR DESCRIPTION
**Pull request by bda618**
## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/ldap.py [hunk 1]
```diff
@@ -374,6 +374,16 @@ def build_corp_user_mce(
             if self.config.user_attrs_map["countryCode"] in attrs
             else None
         )
+        account_type = (
+            (attrs[self.config.user_attrs_map["accountType"]][0]).decode()
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/ldap.py:378]: How universal is this field in LDAP servers? Do both Microsoft Active Directory and OpenLDAP support it?

File: metadata-ingestion/src/datahub/ingestion/source/ldap.py [hunk 2]
```diff
@@ -374,6 +374,16 @@ def build_corp_user_mce(
...
+            if self.config.user_attrs_map["accountType"] in attrs
+            else None
+        )
+        employee_id = (
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/ldap.py:382]: How universal is this field in LDAP servers? Do both Microsoft Active Directory and OpenLDAP support it?

**Review by jjoyce0510 — COMMENTED** [review]
Hi there! Thanks for the contribution!

I've taken some time to digest this PR. My primary concerns are the following:
1.  account type and employee id, may not be "universal", in the sense that not all LDAP servers provide them.
2. neither of these concept are very relevant to DataHub itself. DataHub cares about the things it needs to depend on (like name, description, email, etc) 

One alternative to adding them as strongly-typed fields across the entire stack (as part of CorpUser models) is to instead introduce a key-value map on the CorpUserInfo.pdl aspect (see customProperties on DatasetProperties.pdl for an example) and then populate these properties with any arbitrary information that you may need. 

We could then render the properties in the UI on the entity profile page (perhaps at the bottom of the left sidebar) 

This would enable us to continue adding new properties that are not DataHub-specific (like these) as new requirements involve.

In short, my recommendation would be:

1. Add string-string map into CorpUserInfo.pdl in optional field called "customProperties" 
3. Produce customProperties containing account type and employee id
4. Wire up the customProperties map all the way to GraphQL layer
5. Render the custom properties in a generic table on the Corp user profile page -> This can be a separate PR for ease of review.

What do you think?

Cheers
John

File: metadata-ingestion/src/datahub/ingestion/source/ldap.py (hunk 2 above)
Comment (bda618) [reply metadata-ingestion/src/datahub/ingestion/source/ldap.py:384]: employeeID appears to be a pretty standart, so I added a default mapping for this ldap attribute into ldap.py as user_attrs_map["employeeId"] = "employeeID". I think this should be good enough for this attribute.

File: metadata-ingestion/src/datahub/ingestion/source/ldap.py (hunk 1 above)
Comment (bda618) [reply metadata-ingestion/src/datahub/ingestion/source/ldap.py:380]: There is no such thing as accountType at ldap,  however, I have found ldap attribute called "employeeType". Now, I am thinking that field name "accountType" as a field for corpUser entity is not a best match and "employeeType" looks better for all others. The thing is that at our enterprise "employeeType" has a different values and doesn't carry the value we need so we don't use one and this is why we came up with name "accountType".

File: metadata-ingestion/src/datahub/ingestion/source/ldap.py (hunk 2 above)
Comment (bda618) [reply metadata-ingestion/src/datahub/ingestion/source/ldap.py:384]: Employee-ID is there: https://learn.microsoft.com/en-us/windows/win32/adschema/a-employeeid
The existing corpUser entity has field "departmentId", so I think "employeeId" is logically matches the place.

I agree that "accountType" is different and should not be there. Instead "employeeType" should be used, it is also standard: https://www.manageengine.com/products/ad-manager/ad-user-attributes/employeetype.html.

EmployeeType usually tells if employee account is for Employee or Contractor.

**Comment by bda618** [comment]
>One alternative to adding them as strongly-typed fields across the entire stack (as part of CorpUser models) is to 
> instead introduce a key-value map on the CorpUserInfo.pdl aspect (see customProperties on DatasetProperties.pdl for > an example) and then populate these properties with any arbitrary information that you may need.
We need to populate this during an ldap ingestion and the current implementation of ldap.py populates only CorpUserInfo, it doesn't populate DatasetProperties. So, essentually an MCE which is generated at ldap.py has no information about anything beyond what was populated into CorpUserInfo.

File: metadata-ingestion/src/datahub/ingestion/source/ldap.py (hunk 2 above)
Comment (bda618) [reply metadata-ingestion/src/datahub/ingestion/source/ldap.py:384]: @jjoyce0510 There is another concern we have just found. It looks like fields at customProperties will not be searchable and we need to commit a search over both fields above. Is there any way around this?

Comment (jjoyce0510) [reply metadata-ingestion/src/datahub/ingestion/source/ldap.py:384]: I believe you can make CustomProperties searchable: https://github.com/datahub-project/datahub/blob/master/metadata-models/src/main/pegasus/com/linkedin/common/CustomProperties.pdl#L15

Since they have the "@Searchable" annotation!

Did you reuse this model?

**Comment by jjoyce0510** [comment]
Ping on this one. I think the team is hesitant to continue to grow the set of typed properties on this model without clear value add on each.

**Comment by bda618** [comment]
> Ping on this one. I think the team is hesitant to continue to grow the set of typed properties on this model without clear value add on each.

@jjoyce0510, I originally discussed this on slack (Dmitry Bryazgin, ingestion topic), but I can provide some details here. A corporate account at enterprise is not always reference to a particular employee, it could be a so called a "service" accounts. Service accounts are usually responsible to run = all kind of scheduled jobs. In fact, ~ 20% of all ldap accounts at our enterprise are "service" accounts. We need a way to distinguish "service" accounts from all others at datahub because they follow different business rules. This is why we wanted to add a new field called "accountType" to the corpUser entity. However, it is clear now that "accountType" is not a standart ldap attribute and nobody else will benefit of this even if they want to have a similar functionality in place. Thus I found a standart ldap field called "employeeType" which could be used as a default mapping. In our case I will remap this to the ldap field that carry the real information we need, all other can have a default value being populated. The same approach could go to employeeId which could be mapped to a standart ldap attribute  employeeID. 

If this approach doesn't work, we can go with proposed solution if custom fields are searchable (I have not verified this yet). I have already extended CorpUserInfo by customProperties and added code to populate this custom properties at ldap.py. I was need to add an extra property to ldap ingestion configuration to list all my custom properties, in this case something like this: custom_prop_list: ["employeeId","accountType"]. I need this property because I don't want to hardcode anything specific at ldap.py and to continue to keep it general. If I have a property, I can easily read all custom fields from there and populate MCE correctly.

The next thing left is "Wire up the customProperties map all the way to GraphQL layer". This is where I will need to investigate more because I am not quite clear how to make this all dynamic (if this even possible and needed).  Please, advise if I miss something.

**Comment by bda618** [comment]
I am closing this PR to create a new one where the above functionality will be implemented through custom properties.
